bee-node/bee-plugin/src/handshake.rs

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Handshake definition for the bee plugin system.
+
+use crate::event::{EventId, InvalidEventId};
+
+use thiserror::Error;
+
+use std::{
+    convert::TryFrom,
+    fmt::Write,
+    net::{AddrParseError, SocketAddr},
+};
+
+/// Information provided by the plugin during the handshake stage.
+pub struct HandshakeInfo {
+    pub(crate) address: SocketAddr,
+    pub(crate) name: String,
+    pub(crate) event_ids: Vec<EventId>,
+}
+
+impl HandshakeInfo {
+    /// Creates a new [`HandshakeInfo`] using the plugin's gRPC server address, the plugin's name for
+    /// logging purposes and the `EventId`s that the plugins will be suscribed to.
+    pub fn new(address: SocketAddr, name: &str, event_ids: Vec<EventId>) -> Self {
+        Self {
+            address,
+            name: name.to_owned(),
+            event_ids,
+        }
+    }
+
+    pub(crate) fn parse(buf: &str) -> Result<Self, InvalidHandshake> {
+        let mut chunks = buf.trim().split('|');
+        let address_chunk = chunks.next().ok_or(InvalidHandshake::MissingAddress)?;
+        let address: SocketAddr = address_chunk.parse()?;
+        let name = chunks.next().ok_or(InvalidHandshake::MissingName)?.to_string();
+        let mut event_ids = vec![];
+
+        for chunk in chunks {
```

Comment:
Can this be reworked in an iter/collect ?

Reply:
yes